snap/squashfs/squashfs.go

```diff
@@ -117,6 +118,99 @@ func (s *Snap) ReadFile(filePath string) (content []byte, err error) {
 	return ioutil.ReadFile(filepath.Join(unpackDir, filePath))
 }
 
+// skippper could also be a map[string]bool, but _as it's only supposed to be
+// checkd through its Has method_, the small added complexity of it being a
+// map[string]struct{} lose to the associated space savings.
+type skipper map[string]struct{}
+
+func (sk skipper) Add(path string) {
+	sk[filepath.Clean(path)] = struct{}{}
+}
+
+func (sk skipper) Has(path string) bool {
+	for p := filepath.Clean(path); p != "." && p != "/"; p = filepath.Dir(p) {
+		if _, ok := sk[p]; ok {
+			return true
+		}
+	}
+
+	return false
+}
+
+// Stat returns an os.FileInfo of a single file or directory inside a squashfs snap.
+//
+// The Sys() method of the FileInfo returns a SnapFileOwner
+func (s *Snap) Walk(relative string, walkFn filepath.WalkFunc) error {
+	relative = filepath.Clean(relative)
+	if relative == "" || relative == "/" {
+		relative = "."
+	} else if relative[0] == '/' {
+		// I said relative, darn it :-)
+		relative = relative[1:]
+	}
+
+	var cmd *exec.Cmd
+	if relative == "." {
+		cmd = exec.Command("unsquashfs", "-no-progress", "-dest", ".", "-ll", s.path)
+	} else {
+		cmd = exec.Command("unsquashfs", "-no-progress", "-dest", ".", "-ll", s.path, relative)
+	}
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return walkFn(relative, nil, err)
+	}
+	if err := cmd.Start(); err != nil {
+		return walkFn(relative, nil, err)
+	}
+	defer cmd.Process.Kill()
+
+	scanner := bufio.NewScanner(stdout)
+	// skip the header
+	for scanner.Scan() {
+		if len(scanner.Bytes()) == 0 {
+			break
+		}
+	}
+
+	skipper := make(skipper)
+	for scanner.Scan() {
+		st, err := fromRaw(scanner.Bytes())
+		if err != nil {
+			err = walkFn(relative, nil, err)
+			if err != nil {
+				return err
+			}
+		} else {
+			path := filepath.Join(relative, st.Path())
+			if skipper.Has(path) {
+				continue
+			}
+			err = walkFn(path, st, nil)
```

Comment:
This looks slightly strange, maybe subjective but it seems something like:
```
                      err = walkFn(path, st, nil) 
                      if err != nil {
				if err == filepath.SkipDir && st.IsDir() {
					skipper.Add(path)
				} else {
					return err
				}
			}
```
is easier to read (i.e. more easy to scan linearly).

Reply:
perfect